### DISCUSSION TITLE
Type Management system prerelease testing

### DISCUSSION THREAD

**Discussion by Frooxius** [Yellow-Dog-Man/Resonite-Issues#2490] [Pre-release Testing]
Hello everyone!

We're in need of yet another prerelease test!

## Build Version
TESTING IS OVER

## BEFORE YOU RUN THE BUILD
The latest test builds break compatibility! Anything saved with those will not load on the main branch anymore - this includes Local, Dash and Settings.

I strongly recommend adding `-NeverSaveSettings` and `-NeverSaveDash` as launch arguments before testing further, otherwise you might LOSE YOUR SETTINGS and LOSE YOUR DASH if you go back to the normal build!

## What is being tested
This prerelease features a new "Type Management" system. This fundamentally changes how data types are loaded/encoded and transmitted on the network, in order to improve compatibility and security and a necessary step to transition towards modern .NET runtime.

Since this changes how data is decoded and encoded, there's potential for content breakage - existing content might not load properly in some cases. While I've done a bunch of testing myself, I'd like to have wider community verify those changes.

If you'd like to know details, check out this issue for more: https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/2264

## How to test
- On Steam, switch to the `prerelease` branch and make sure the build updates before you run it
- Make sure your build version matches the one at the top of this message

You can test either alone or with multiple users (which helps test networking aspect). Even testing alone by loading a lot of content will help.

- Load saved worlds and items
-- Any worlds/items work! Yours, published ones, shared folder
-- Try to go through range of content - from recent, to older and really old legacy content that hasn't been updated
- Check if the worlds/items function properly or if there's sudden breakage that is not present on the main built - e.g. missing components, or cleared out fields that shouldn't be
- If testing with multiple users, have them join the sessions you host to see if networking transmits everything properly
- If you're joining other users, make sure you're not forcing connections through relay - we want to test direct connections where possible!
- Save copies of items & worlds and see if they load properly

## What NOT to test & report
- World loading time - this will also be addressed before release - the new system is currently missing caching and has some extra diagnostics, which slows things down a fair bit
- Types that aren't normally expected to be used - e.g. dynamic variables of "FrooxEngineRunner" type and similar types that do not have any practical use. If you're unsure, report it anyway and we'll evaluate if given type is expected to break or not.

## If you find something that got broken
- Switch to the normal build and try to load the same world/item to verify that the issue is present ONLY on the prerelease build
- Create a GitHub issue for a prerelease bug: https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/new?assignees=shiftyscales&labels=prerelease&projects=&template=prerelease.yml
- Make sure to include the log (you can also check it for messages stating which types failed to load)
- Please include description of what happened. Be succinct, but exact.
- **DO NOT SAVE** your worlds and other items - they might save in a broken state and be unrecoverable

## If you find that stuff works okay
Please report it too and include details on what you generally tested, so we know things are okay. You don't need to make a new issue for this - just a commend in this Discussion is fine.

## How long will testing go for?
With the nature of the changes, we expect the issues to become apparent relatively quickly, so unlikely more than a few days. Once we get a good feel that there's no breakage, this will get merged to main build.

**Comment by epicEaston197** [comment]
Hey I'm testing the new prerelease and I wanted to edit one of my settings
Just to find out that the settings menu does not exist
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/76523692/dfcea24e-2d87-4fb4-9546-966a7088f205)

**Reply by ModernBalloonie** [reply]
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/82356528/cf3b7276-810a-4f16-9776-8c0101f3a670)
Same here.

**Reply by Frooxius** [reply]
Please follow the instructions in the post when reporting stuff. We ask that you:
- Create an issue instead of posting
- Include logs

**Reply by coolymike** [reply]
Same issue on the Linux native build.
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/11460382/6f61663f-90cb-483d-b44c-054e11b26990)

**Reply by bredo228** [reply]
Seeing the same issue.

![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/29470709/2727a7df-5751-46ca-97d4-4208a38432bd)

https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/2491

**Comment by Frooxius** [comment]
I've updated the pre-release build! Copying release notes. Going forward, make sure to test with this version (or newer if one is released by then).

2024.7.8.1130 [PRE-RELEASE]
This is a pre-release build for the new type management system. Thanks to everyone reporting issues, this should fix most of them.

If you'd like to help with testing, check out this GitHub Discussion and please read the instructions: https://github.com/Yellow-Dog-Man/Resonite-Issues/discussions/2490

**New Features:**
- Added type validation when assigning types to type fields

**Bugfixes:**
- Mark data feed items as data model types (based on report by @bredo, @epiceaston197, @modernballoonie, @whippersnatchpumpkinpatch, issue #2491)
- Fixed certain settings (LeapMotion & Windows Settings) not being properly included in the data model types (based on report by @epiceaston197, issue #2491)
- Mark dash screen initializers as data model types (based on report by @bredo, @epiceaston197, @modernballoonie, @whippersnatchpumpkinpatch, @zyro1331, issue #2491)
- Mark ButtonEventData as data model type to fix exceptions and crashes in some cases when opening context menu (based on report by @modernballoonie, @epiceaston197 and @bredo, issue #2493)
- Properly map nested types when their parent type is moved
- Mark additional system types as data model types (based on report by @whippersnatchpumpkinpatch, gameboyjp, issue #2498)
- Fixed mapping for a number of legacy types (based on report by @whippersnatchpumpkinpatch, issue #2499)
- Fix component selector breaking with generic types (based on report by @whippersnatchpumpkinpatch, issue #2492)

**Comment by Frooxius** [comment]
I've updated prerelease build again! Copying release notes here:

**2024.7.8.1285 [PRE-RELEASE]**
Another pre-release build for the new type management system. This fixes a number of other issues that have been found. This should fix most of the additional issues that were found.

If you'd like to help with testing, check out this GitHub Discussion and please read the instructions: https://github.com/Yellow-Dog-Man/Resonite-Issues/discussions/2490

**New Features:**
- Add data type validation when assigning targets to sync delegate
-- This will help catch some type compatibility issues quicker and prevent corrupting the data model with unsupported types

**Optimize:**
- Added caching to type decoding to speed up decoding process some

**Fixes:**
- Add color32 to data model types
-- This fixes crashes with StaticTexture2D and other things (reported by @zyro1331, @epiceaston197, issue #2503)
- Fixed opening context menus with cancel option resulting in a crash (reported by @modernballoonie, @epiceaston197, @bredo, @zyro1331, issue #2493)
- Fix binary type encoding not being able to distinguish generic type definition from generic type instance
-- This fixes crash when opening generic components in the generic component browser (reported by @zyro1331, @epiceaston197, issue #2505)
- Fix moved type remapping not handling cases where the moved type is a generic type, but the old typename doesn't specify a generic marker
-- This fixes some data types like ContinouslyChangingObjectRelay not loading properly (reported by @whippersnatchpumpkinpatch, issue #2506)
- Fix a number of ProtoFlux datatypes not being included (based on report by @bredo, issue #2508)

**Comment by Frooxius** [comment]
Another pre-release build! Please update this one!

**2024.7.9.1044 [PRE-RELEASE]**
Addressing more issues found with the new type management system on the pre-release. This is not a normal build, it's for testing purposes only!

If you'd like to help with testing, check out this GitHub Discussion and please read the instructions: https://github.com/Yellow-Dog-Man/Resonite-Issues/discussions/2490

**New Features:**
- Expand type validation to validate all generic arguments on base classes of data model types
-- This helps catch a number more types that might end up being invalid in the data model

**Tweaks:**
- Do not generate bindings for DSP nodes at the moment, since it's fully implemented yet
- Consider any types that implement data model interfaces to also be implicitly data model types

**Bugfixes:**
- Mark a batch of types as data model types to fix a number of issues
-- This fixes texture references not spawning properly (reported by @epiceaston197, @zyro1331)
- SyncType editor now uses the data model type parsing
-- This ensures that invalid types cannot be referenced in the data model
-- This also fixes needing to specify the full namespace when typing types (based on report by @bluecyro, issue #2511)
- Fix types with no namespaces formatting incorrectly (based on report by Nihlus, issue #2504)

**Comment by Zyro1331** [comment]
I've been testing Beta 2024.7.9.1044 with a public session for over an hour, we have not been able to identify any issues while doing some basic content creation. 
The only possible problems I've noticed is the really large lag spikes upon loading the contacts or worlds menu, which are noticeably higher than the Stable build. My client will lock up for about ~10 full seconds upon opening the contacts list, instead of the usual ~3 _(complete "guestimate" numbers, I haven't actually timed this or done speed comparisons)_ 

Another identified problem _(this also affects stable build as of today)_ is yt-dlp not loading YouTube videos, but upon doing a manual update of the yt-dlp library by using `yt-dlp -U` via CLI fixed this problem.

**Reply by Frooxius** [reply]
Thanks for update, but for any issues that also affect stable build please leave them out - we'll deal with those separately.

**Reply by ohzee00** [reply]
I also noticed the increased stutters and wanted to make a issue about it but the "what NOT to report" does go over worlds loading time and the like, which I can assume can affect those dash elements too in the prereleases current uncached state.

**Comment by Frooxius** [comment]
EDIT: 2024.7.9.1422 is out of date, I've pushed a small change. We still want to test the same things mentioned for this build, but READ THE LATEST BUILD NOTES before you proceed!

**2024.7.9.1422 [PRE-RELEASE]**
Another pre-release! The decoding seems to be pretty stable now, so I've implemented most of the remaining major things for the new type management system - actually using the new encode system for types when saving items!

For anyone testing: If you can - please run through a bunch of things again, since this rejiggled how loading works, so it's possible there are some regressions again. Try new content, old content. Save copies of items and worlds too and see if those load properly as well.

**IMPORTANT**: Be EXTRA careful saving stuff on this build! It won't load on the main one! This includes dash, settings & local! You might need to reset your dash after you go back on normal. 

There's also caching of some of the metadata that needs to be computed, so startup is much faster now (unless some of the libraries changed, but we'll usually have the vanilla data pre-computed with each release).

**New Features:**
- Switch type encoding to use the new system & rework how type encoding is structured for better efficiency
-- This mostly "finalizes" the type management system, switching over to the new system fully
-- This fixes a number of issues with type compatibility for SyncDelegates, SyncType and SyncVar fields
-- Types are now only encoded once for each unique type in a type table, which speeds up both encoding and decoding
-- IMPORTANT: Anything you have with this version won't be able to be loaded on older builds. Save carefully! (this includes implicitly saved things like the dash or the settings!)

**Optimize:**
- Implemented mechanism to cache some assembly metadata
-- This will heavily speed up start up time when the metadata is cached as it doesn't need to be computed on every startup

**Reply by Frooxius** [reply]
Copied wrong build number, it's 2024.7.9.1422

**Reply by Frooxius** [reply]
Also I'm not kidding about being careful with this build. This is what I see after going back to the normal branch: 
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/8838625/69298a88-6c48-4909-9703-e87e94995b79)

**Reply by Frooxius** [reply]
I've published another extra build which helps. I STRONGLY RECOMMEND READING BELOW AND USING THIS BUILD:

**2024.7.10.29 [PRE-RELEASE]**
Another pre-release build! This is just quick quality of life addition for testing, which prevents settings & dash from being saved with commandline arguments.

If you are helping to test the latest build (see the previous update above), I STRONGLY RECOMMEND you use these arguments, otherwise you might lose your dash and settings, which will be harder to recover. 

If you'd like to help with testing, check out this GitHub Discussion and please read the instructions: https://github.com/Yellow-Dog-Man/Resonite-Issues/discussions/2490

**New Features:**
- Added `-NeverSaveSettings` and `-NeverSaveDash` launch arguments to prevent settings/dash from being saved on the build
-- This is particularly useful for testing purposes to not end up corrupting / breaking these

**Reply by Frooxius** [reply]
Another build. I'm keeping this whole chain of  builds open, since we're still testing the things specified in the build above, these are more adding extra polish.

**2024.7.10.903 [PRE-RELEASE]**
Another pre-release build testing the new type system. This is mostly optimizations, removing the old systems that were replaced by the new and such. I don't see any issues cropping up, so I think this is pretty close to getting finished. I'll wrap up a few more of the issues and leave the testing open for a bit more before we pull this into main.

If you'd like to help with testing, check out this GitHub Discussion and please read the instructions: https://github.com/Yellow-Dog-Man/Resonite-Issues/discussions/2490

I STRONGLY RECOMMEND you read the instructions in the testing link above, otherwise you might LOSE YOUR DASH/SETTINGS and other stuff!

**Optimizations:**
- Added caching for determining if types are supported or not, to speed up validation operations
- Cache session compatibility information, to avoid having to compute the compatibility for the same set of assemblies every time
- Removed old type remapping system

**Reply by Frooxius** [reply]
**2024.7.10.1077 [PRE-RELEASE]**
Another pre-release build! This is polishing some more bits so types with arrays are formatted and parsed properly. This is mostly internal things - this doesn't mean that you can use arrays or anything like that, but fixes instances where their internal use leads to wrong results when encoding types.

I'm now considering wrapping this up and merging this soon, since it's pretty quiet on more issues! If you'd like to help do last round of testing to see if there's any more issues check out this GitHub Discussion and please read the instructions: https://github.com/Yellow-Dog-Man/Resonite-Issues/discussions/2490

I STRONGLY RECOMMEND you read the instructions in the testing link above, otherwise you might LOSE YOUR DASH/SETTINGS and other stuff!

**New Features:**
- Add support for properly formatting array types (based on report by @jgullberg, issue #2504)
-- This includes any combination of array ranks and arrays of arrays and combinations of thereof
- Add support for parsing array types with custom nice type parser (based on report by @jgullberg, issue #2504)
-- This supports arrays of arrays and arrays of any rank and any combinations of thereof

**Reply by Frooxius** [reply]
**2024.7.10.1266 [PRE-RELEASE]**
Another pre-release update! This is just trickling in some smaller reports now, this has a small bugfix. 

There don't seem to be any big major blockers so far, so this is likely to get merged today. If you'd like to help do last round of testing to see if there's any more issues check out this GitHub Discussion and please read the instructions: https://github.com/Yellow-Dog-Man/Resonite-Issues/discussions/2490

I STRONGLY RECOMMEND you read the instructions in the testing link above, otherwise you might LOSE YOUR DASH/SETTINGS and other stuff!

**Bugfixes:**
- Writing unsupported types into DataModelTypeStore will now trigger fail, instead of an exception (based on report by @mpmxyz, issue #2522)

**Reply by Frooxius** [reply]
**2024.7.10.1338 [PRE-RELEASE]**
Another pre-release! I think things are looking good to merge now, so I'll start prepping the build. Thanks everyone for helping to test!

If you'd like to still test while I prep other things, check out this GitHub Discussion and please read the instructions: https://github.com/Yellow-Dog-Man/Resonite-Issues/discussions/2490

I STRONGLY RECOMMEND you read the instructions in the testing link above, otherwise you might LOSE YOUR DASH/SETTINGS and other stuff!

**Bugfixes:**
- Fix writing dynamic type variables throwing an exception when the type value is not supported data model type (based on report by @mpmxyz, issue #2522)

**Reply by Frooxius** [reply]
**2024.7.11.377 [PRE-RELEASE]**
Another small pre-release update, fixing up another issue. I'll have to deal with the big one tomorrow and the rest, but this should fixup a thing.

If you'd like to still test while I prep other things, check out this GitHub Discussion and please read the instructions: https://github.com/Yellow-Dog-Man/Resonite-Issues/discussions/2490

I STRONGLY RECOMMEND you read the instructions in the testing link above, otherwise you might LOSE YOUR DASH/SETTINGS and other stuff!

**New Features:**
- Added encoding/decoding for quantity types, allowing them to be used as values in the data model
-- This fixes crashes when trying to create fields/stores of these types (reported by @_deltawolf, @aritosuits and @stiefeljackal, issue #2527)

**Notes:**
- The latest changes from main build were merged with pre-release and has the latest updates from the pre-release build

**Reply by Frooxius** [reply]
**2024.7.11.1165 [PRE-RELEASE]**
Fixing up the last found major issue with the new type management system! Graphical client and Headless are now properly compatible with each other and produce same hashes for same assemblies. This also fixes up a few more smaller issues as well.

I think this is now pretty much in a state where it's ready to be merged and released. I'll leave the testing open a bit more while I tackle some other stuff, so if you'd like to hop on a verify these yourself, now's your chance, but I don't see any other major issues remaining at this point. Check out this GitHub Discussion and please read the instructions: https://github.com/Yellow-Dog-Man/Resonite-Issues/discussions/2490

I STRONGLY RECOMMEND you read the instructions in the testing link above, otherwise you might LOSE YOUR DASH/SETTINGS and other stuff!

**Tweaks:**
- Component selector will not list any components that aren't allowed in the current session
- Add `void` as alias for the void type
- Add `void` as system type

**Fixes:**
- Fix different enumeration order on methods on workers between different runtimes resulting in different compatibility hash when it should be the same
- Ensure that headless finds all necessary data model assemblies when scanning them
- Expand data model assembly categorization to include userspace core assemblies, that are only available in userspace
-- This is for certain setting components, that will only exist in the userspace, but cannot exist in the world space
-- This fixes headless and graphical client being incompatible due to graphical client having additional components that headless doesn't
- Fix nested types used as generic arguments not parsing properly
-- This fixes not being able to select nested types in component/protoflux selector as generics arguments (reported by @mpmxyz and @_deltawolf, issue #2525)
- ComponentSelector will not list common generic variants that aren't valid data model types for given session

**Reply by TisFoolish** [reply]
Does this lay the groundwork for allowing clients and headlesses of different versions being able to connect to each other?

**Reply by Frooxius** [reply]
Depends on what you mean by that exactly. 

If they're different builds, but they're compatible, they'll be able to connect to each other - but that was already the case. The main difference there is that this is now computed automatically, rather than us having to manually mark them as incompatible builds.

If they're not compatible, then they won't be able to connect - incompatible builds/assemblies will never be able to connect to each other.

However it does lay some important groundwork for being able to run several versions in parallel, so you can connect to different versions, without having to restart, but that itself requires additional work.

**Reply by TisFoolish** [reply]
> However it does lay some important groundwork for being able to run several versions in parallel, so you can connect to different versions, without having to restart, but that itself requires additional work.

This is what I meant and glad to see that confirmed

**Comment by TisFoolish** [comment]
How long will people have time to test before its pushed to the wider public

**Reply by Frooxius** [reply]
Probably a few more hours. I'm wrapping a few other things and working on some other stuff in the meanwhile. Then I'll see what's the status of things and see if we're good to merge or not.

**Reply by TisFoolish** [reply]
That's not a lot of time for people to test. A lot of the US testers are likely at work still, who seem to be most of the testers atm.

**Reply by Zyro1331** [reply]
I have been testing for a few hours between different prerelease versions, saving loading content from various builds, doing content creation and editing, I have not identified any issues other than minor slowdowns in some parts.

**Reply by Frooxius** [reply]
We've been testing for a few days already and caught the things we needed to.

The last report couldn't find any more major issues even after a while of testing, so there's no longer a big concern that things are significantly broken. I've also been doing focused testing throughout, so I'm pretty confident at this point there's nothing major left.

Right now things are in a state where it's just re-checking things to "rubber stamp" it.

**Comment by ohzee00** [comment]
Did a quick test on the newest version (2024.7.10.1077) and the things I tested were:

- Going around various worlds and creations
- Exporting a Resonite package and importing it
- Saving said Resonite package and spawning it out
- Saving a world and opening it
- "Wiping" my database and cache and going through the new player experience without creating a account (I did open the create account dialog though just to see if it was still there)

In those quick tests I have found nothing wrong and everything seems to be working as it should, granted my tests are not as intense as others and something could fall through the cracks however to me, this seems pretty solid. I did not test with others, that is one thing I can't easily do currently.

The only iffy thing I could mention is that the Worlds and Contacts dash sections _feel_ like they take much longer to load the first time, but that could be placebo, long stutters are difficult to get accurate numbers for.

**Comment by Frooxius** [comment]
Testing is now over and this is out! Thank you everyone!
